### ISSUE TITLE
Check for filter_map combined with `bool::then`, suggest `filter` then `map`

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/rust-clippy#9098]
### What it does

The combination of `filter_map` with `bool::then` can often be more readably written as a `filter` followed by a `map`, which avoids multiple levels of closure nesting and reduces indentation.

### Lint Name

filter_then_map

### Category

style

### Advantage

- Avoids multiple levels of closure and parenthesis nesting.
- Reduces indentation.

### Drawbacks

- While the compiler should generally be able to optimize `filter` then `map`, this may potentially miss optimization opportunities.

### Example

```rust
let v = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30];
collection
    .iter()
    .filter_map(|n| (n % 2 == 0).then(|| complex_calculation(n * 100)))
```

Could be written as:

```rust
let v = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30];
collection
    .iter()
    .filter(|n| n % 2 == 0)
    .map(|n| complex_calculation(n * 100))
```

(The case I ran into that prompted this had a dozen lines in place of `complex_calculation`.)

**Comment by Centri3** [comment]
The two actually produce [identical assembly](https://rust.godbolt.org/z/xG1s3K4KE).

@rustbot claim
